Add TimeProgressEstimator class

jyavenard opened the pull request:
#### f9a82eb37c5e5616212cd4f208bb17b6cc45cc2f
<pre>
Add TimeProgressEstimator class
<a href="https://bugs.webkit.org/show_bug.cgi?id=268911">https://bugs.webkit.org/show_bug.cgi?id=268911</a>
<a href="https://rdar.apple.com/122463053">rdar://122463053</a>

Reviewed by Eric Carlson and Youenn Fablet.

Simplify and move the emulated time progress management to its own class.
Fly-by: the time would have continued to progress when pause() was called
(until we got a new time update from the GPU process), reading the time following
a call to pause() would have temporarily returned the wrong value.

When seeking, the time should have stopped until the GPU process signalled we
had completed the seek.

The class is thread-safe when used with MediaSource, when used with CRABS
it will only be called from the main thread.

Covered by existing tests.

* Source/WebKit/WebProcess/GPU/media/MediaPlayerPrivateRemote.cpp:
(WebKit::MediaPlayerPrivateRemote::TimeProgressEstimator::TimeProgressEstimator):
(WebKit::MediaPlayerPrivateRemote::TimeProgressEstimator::currentTime const):
(WebKit::MediaPlayerPrivateRemote::TimeProgressEstimator::cachedTime const):
(WebKit::MediaPlayerPrivateRemote::TimeProgressEstimator::timeIsProgressing const):
(WebKit::MediaPlayerPrivateRemote::TimeProgressEstimator::pause):
(WebKit::MediaPlayerPrivateRemote::TimeProgressEstimator::setTime):
(WebKit::MediaPlayerPrivateRemote::TimeProgressEstimator::setRate):
(WebKit::MediaPlayerPrivateRemote::MediaPlayerPrivateRemote):
(WebKit::MediaPlayerPrivateRemote::pause):
(WebKit::MediaPlayerPrivateRemote::durationMediaTime const):
(WebKit::MediaPlayerPrivateRemote::currentMediaTime const):
(WebKit::MediaPlayerPrivateRemote::seekToTarget):
(WebKit::MediaPlayerPrivateRemote::seeked):
(WebKit::MediaPlayerPrivateRemote::rateChanged):
(WebKit::MediaPlayerPrivateRemote::playbackStateChanged):
(WebKit::MediaPlayerPrivateRemote::currentTimeChanged):
(WebKit::MediaPlayerPrivateRemote::performTaskAtMediaTime): fly-by: completionHandler wasn&apos;t called under some circumstances.
* Source/WebKit/WebProcess/GPU/media/MediaPlayerPrivateRemote.h:

Canonical link: <a href="https://commits.webkit.org/274278@main">https://commits.webkit.org/274278@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/69268e972cbd2be755686342d6c1f981099fdd5f

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/38441 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/17385 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/40780 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/40992 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/34160 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/20162 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/14729 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/32366 "Found 1 new test failure: media/track/media-element-enqueue-event-crash.html (failure)") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/39014 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/14665 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/33482 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/12745 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/12726 "Passed tests") | | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/42270 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/34949 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/34736 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/38564 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/13323 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/11022 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/36766 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/14901 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/8649 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/13791 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/14373 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebKit/WebProcess/GPU/media/MediaPlayerPrivateRemote.h:
```diff
@@ -197,6 +198,25 @@ class MediaPlayerPrivateRemote final
 private:
+    class TimeProgressEstimator final {
+    public:
+        TimeProgressEstimator(const MediaPlayerPrivateRemote& parent);
```
explicit

youennf commented on Source/WebKit/WebProcess/GPU/media/MediaPlayerPrivateRemote.cpp:
```diff
@@ -401,15 +447,15 @@ bool MediaPlayerPrivateRemote::seeking() const
 void MediaPlayerPrivateRemote::rateChanged(double rate)
 {
     m_rate = rate;
+    m_currentTimeEstimator.setRate(rate);
```
Does it make sense to duplicate m_rate in MediaPlayerPrivateRemote and time estimator?

eric-carlson replied:
Yes, it seems wrong to cache rate and query `m_parent.durationMediaTime`. Should either cache both or none.

jyavenard replied:
I did so because one is only ever accessed on the main thread, while the other requires a lock.

youennf commented on Source/WebKit/WebProcess/GPU/media/MediaPlayerPrivateRemote.cpp:
```diff
@@ -110,14 +111,67 @@ using namespace WebCore;
+    m_timeIsProgressing = false;
+}
+
+void MediaPlayerPrivateRemote::TimeProgressEstimator::resync(const MediaTime& time, const MonotonicTime& wallTime, std::optional<bool> timeIsProgressing)
```
setTimes instead of resync?

jyavenard committed:
Add TimeProgressEstimator class
https://bugs.webkit.org/show_bug.cgi?id=268911
rdar://122463053

Reviewed by Eric Carlson and Youenn Fablet.

Simplify and move the emulated time progress management to its own class.
Fly-by: the time would have continued to progress when pause() was called
(until we got a new time update from the GPU process), reading the time following
a call to pause() would have temporarily returned the wrong value.

When seeking, the time should have stopped until the GPU process signalled we
had completed the seek.

The class is thread-safe when used with MediaSource, when used with CRABS
it will only be called from the main thread.

Covered by existing tests.

* Source/WebKit/WebProcess/GPU/media/MediaPlayerPrivateRemote.cpp:
(WebKit::MediaPlayerPrivateRemote::TimeProgressEstimator::TimeProgressEstimator):
(WebKit::MediaPlayerPrivateRemote::TimeProgressEstimator::currentTime const):
(WebKit::MediaPlayerPrivateRemote::TimeProgressEstimator::cachedTime const):
(WebKit::MediaPlayerPrivateRemote::TimeProgressEstimator::timeIsProgressing const):
(WebKit::MediaPlayerPrivateRemote::TimeProgressEstimator::pause):
(WebKit::MediaPlayerPrivateRemote::TimeProgressEstimator::setTime):
(WebKit::MediaPlayerPrivateRemote::TimeProgressEstimator::setRate):
(WebKit::MediaPlayerPrivateRemote::MediaPlayerPrivateRemote):
(WebKit::MediaPlayerPrivateRemote::pause):
(WebKit::MediaPlayerPrivateRemote::durationMediaTime const):
(WebKit::MediaPlayerPrivateRemote::currentMediaTime const):
(WebKit::MediaPlayerPrivateRemote::seekToTarget):
(WebKit::MediaPlayerPrivateRemote::seeked):
(WebKit::MediaPlayerPrivateRemote::rateChanged):
(WebKit::MediaPlayerPrivateRemote::playbackStateChanged):
(WebKit::MediaPlayerPrivateRemote::currentTimeChanged):
(WebKit::MediaPlayerPrivateRemote::performTaskAtMediaTime): fly-by: completionHandler wasn't called under some circumstances.
* Source/WebKit/WebProcess/GPU/media/MediaPlayerPrivateRemote.h:

Canonical link: https://commits.webkit.org/274278@main
